packages/internal/src/build/api.ts

```diff
@@ -27,29 +28,93 @@ export const cleanApiBuild = () => {
   rimraf.sync(path.join(rwjsPaths.generated.prebuild, 'api'))
 }
 
+export const getPrebuildOutputOptions = (
```

Comment:
These are good points, and I totally don't like how this code reads either. I'm not sure I fully understand this yet, but once I start playing around with it maybe it will.

I'll see if I can implement your suggestions - however, this did build on the existing convention in the file before I think i.e. we were getting the path and content inline within the transpliation-step.

Reply:
Ok, so I think I understand the suggestion here now. Thanks for the detailed review Peter! Some stuff to discuss:

> 1. Remove this code from "prebuild", and rather invoke it afterwards in the "buildApi" function.

I can see your logic here - but let try and convince you this is the better place for it. If the purpose of prebuild is to "remove all redwood magic", i.e. leave es6 js files that will just run, I think it makes perfect sense to leave it in prebuild. The way I see it, this step allows functions to run on all our supported providers (netlify, vercel, render, self-hosted) - and they all behave consistently. 

Without generating these proxy functions, what we get is code that won't quite work on all providers, because they do not support nesting consistently. I think it also is neater that the esbuild step does only one thing - transpiles from es6 to the target file type (the function name even says `transpileApi`).


> 2. Filter out the functions that are not important; you're only concerned with subfolder function files, and a nice way of doing this that is nicely testable is adding a isApiSubfolderFunction tester. (see files.ts::isApiFunction)

Just confirming the suggestion here, in pseudo-code, should the suggested implementation look like this?

```js 
buildApi() {
	const nonNestedApiFiles = findApiFilesWithoutNestedFunctions()
	const nestedFunctions = findNestedFiles()

	const allPrebuiltFiles = [...preBuildApiFiles(nonNestedApiFiles), ...preBuildNestedFunction(nestedFunctions)]

	return transpileApi(allPrebuiltFiles)
}
```


 
> 3. Make it do all the work in this single function: iterate over the files, figure out the paths, and write the content.

If we must split the lookup of api files, question here would be there's 2 ways we could group the api files:

- Set 1: 
a) All functions 
b) All other files

or

- Set 2: 
a) Non-nested functions + all other files
b) Nested functions only

Personally I prefer Set 1 - but totally open to other ideas.

Reply:
I hear you, but I disagree.

I think there are two concepts:
1. Transpile the user's RedwoodJS code in standard JavaScript.
2. Supported serverless functions in subdirectories by moving the subdirectories and generating a proxy module.

In a broad sense I agree that they're "enabling redwood magic," but I still think they're doing two separate things to achieve that goal.

Here's a thought experiment:

A year from now we decided that "no on uses serverless functions in subdirectories anymore!" And want to delete it... we would have to scratch around in the transpilation code (that we don't want to remove), and worse yet, the return type is modified, so we would have to check all the places that the transpilation code is invoked to make sure we're doing a good job.

When introducing new functionality I try to determine what it would take to delete that code if the functionality was no longer required. The idea is that deleting code should be much easier than fixing bugs and adding new functionality. If deleting it is difficult, then fixing bugs and adding functionality is going to be _really_ difficult because it's probably too tightly coupled, and requires you to understand more than what the functionality introduces.

Alternatively if you added it in the way I suggested above then deleting it would be removing the invocation of a single function and removing that function. (Also super easy to test.)

Reply:
If you want we can pair and refactor this implementation.

Reply:
Refactored :) - theres is a little down side, which is that I'm having to move files. But if we wanted to undo this, you literally have to remove one line.